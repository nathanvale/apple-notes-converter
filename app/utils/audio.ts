import audioBufferToWav from 'audiobuffer-to-wav'

/**
 * Converts an audio Blob to a mono WAV Blob with a specified sample rate.
 * Suitable for coverting audio for OpenAI's Speech API.
 *
 * @param audioBlob - The input audio Blob to be converted.
 * @param targetSampleRate - The target sample rate for the output WAV file (default is 16,000 Hz).
 * @returns A Promise that resolves to a Blob containing the converted mono WAV audio.
 * @throws Will throw an error if the conversion process fails.
 *
 * @example
 * ```typescript
 * const audioBlob = new Blob([...], { type: 'audio/webm' });
 * const wavBlob = await convertAudioBlobToMonoWav(audioBlob, 16000);
 * ```
 */

export const convertAudioBlobToMonoWav = async (
	audioBlob: Blob,
	targetSampleRate = 16000,
): Promise<Blob> => {
	try {
		// Create an AudioContext with the target sample rate of 16,000 Hz
		const audioContext = new AudioContext()

		// Convert the webm blob to an ArrayBuffer
		const arrayBuffer = await audioBlob.arrayBuffer()

		// Decode the audio data from the arrayBuffer
		const audioBuffer = await audioContext.decodeAudioData(arrayBuffer)

		// Create a new AudioContext for resampling to 16 kHz
		const offlineAudioContext = new OfflineAudioContext(
			1, // Mono output
			(audioBuffer.length * targetSampleRate) / audioBuffer.sampleRate,
			targetSampleRate,
		)

		// Create a buffer source
		const bufferSource = offlineAudioContext.createBufferSource()
		bufferSource.buffer = audioBuffer

		// Connect the buffer source to the destination (i.e., the output)
		bufferSource.connect(offlineAudioContext.destination)
		bufferSource.start(0)

		// Start rendering the audio
		const resampledBuffer = await offlineAudioContext.startRendering()

		// Convert the resampled mono audio buffer to WAV format
		const wavBuffer = audioBufferToWav(resampledBuffer)

		// Create a new Blob with the WAV buffer and correct MIME type
		return new Blob([wavBuffer], { type: 'audio/wav' })
	} catch (error) {
		console.error('Error converting audio blob to a mono wav blob:', error)
		throw error
	}
}
