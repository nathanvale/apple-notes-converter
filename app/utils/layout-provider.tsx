import { createContext, type ReactNode, useContext, useState } from 'react'

interface LayoutContextProps {
	isSideNavOpen: boolean
	toggleNav: () => void
}

const LayoutContext = createContext<LayoutContextProps | undefined>(undefined)

export const LayoutProvider = ({ children }: { children: ReactNode }) => {
	const [isSideNavOpen, setIsSideNavOpen] = useState(true)

	const toggleNav = () => {
		setIsSideNavOpen((prev) => !prev)
	}

	return (
		<LayoutContext.Provider value={{ isSideNavOpen, toggleNav }}>
			{children}
		</LayoutContext.Provider>
	)
}

export const useLayoutContext = () => {
	const context = useContext(LayoutContext)
	if (!context) {
		throw new Error('useLayoutContext must be used within a LayoutProvider')
	}
	return context
}
