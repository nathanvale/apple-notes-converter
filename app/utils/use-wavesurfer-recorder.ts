// useWavesurferRecorder.tsx
import { useWavesurfer } from '@wavesurfer/react'
import { useCallback, useEffect, useReducer, useRef } from 'react'
import RecordPlugin from 'wavesurfer.js/dist/plugins/record.esm.js'

const DEFAULT_WAVE_COLOR = '#737373'
const ERROR_WAVE_COLOR = '#ef4444'

// Define the state interface
interface RecorderState {
	isRecording: boolean
	isProcessing: boolean
	error: string
	secondsElapsed: number
	transcription: string
}

// Define the action types
type RecorderAction =
	| { type: 'START_RECORDING' }
	| { type: 'STOP_RECORDING' }
	| { type: 'SET_PROCESSING'; payload: boolean }
	| { type: 'SET_TRANSCRIPTION'; payload: string }
	| { type: 'SET_ERROR'; payload: string }
	| { type: 'SET_SECONDS_ELAPSED'; payload: number }
	| { type: 'RESET_STATE' }

// Initial state
const initialState: RecorderState = {
	isRecording: false,
	isProcessing: false,
	error: '',
	secondsElapsed: 0,
	transcription: '',
}

// Reducer function
function reducer(state: RecorderState, action: RecorderAction): RecorderState {
	switch (action.type) {
		case 'START_RECORDING':
			return {
				...state,
				isRecording: true,
				error: '',
				secondsElapsed: 0,
			}
		case 'STOP_RECORDING':
			return { ...state, isRecording: false }
		case 'SET_PROCESSING':
			return { ...state, isProcessing: action.payload }
		case 'SET_TRANSCRIPTION':
			return { ...state, transcription: action.payload }
		case 'SET_ERROR':
			return { ...state, error: action.payload }
		case 'SET_SECONDS_ELAPSED':
			return { ...state, secondsElapsed: action.payload }
		case 'RESET_STATE':
			return { ...initialState }
		default:
			return state
	}
}

export function useWavesurferRecorder(
	onRecordingEnd: (blob: Blob) => Promise<void>,
) {
	const waveformRef = useRef<HTMLDivElement | null>(null)
	const [state, dispatch] = useReducer(reducer, initialState)
	const isCancelledRef = useRef(false)
	const recordedBlobRef = useRef<Blob | null>(null)
	const recordPlugin = useRef<RecordPlugin | null>(null)

	const { wavesurfer } = useWavesurfer({
		container: waveformRef,
		waveColor: DEFAULT_WAVE_COLOR,
		progressColor: DEFAULT_WAVE_COLOR,
		cursorColor: 'transparent',
		barWidth: 2,
		height: 24,
		barHeight: 2,
		barGap: 2,
		renderFunction,
		peaks: silentData,
		duration: 0.5,
		backend: 'MediaElement',
		normalize: true,
	})

	const resetState = useCallback(() => {
		dispatch({ type: 'RESET_STATE' })
		isCancelledRef.current = false
		recordedBlobRef.current = null
		wavesurfer?.empty()
		wavesurfer?.setOptions({
			waveColor: DEFAULT_WAVE_COLOR,
		})
	}, [wavesurfer])

	const handleRecordEnd = useCallback(
		async (blob: Blob) => {
			try {
				dispatch({ type: 'SET_PROCESSING', payload: true })
				await onRecordingEnd(blob)
				dispatch({ type: 'SET_PROCESSING', payload: false })
				dispatch({ type: 'STOP_RECORDING' })
			} catch (error) {
				dispatch({ type: 'SET_PROCESSING', payload: false })
				if (error instanceof Error) {
					dispatch({ type: 'SET_ERROR', payload: error.message })
				} else {
					dispatch({
						type: 'SET_ERROR',
						payload: 'An unexpected error occurred',
					})
				}
				wavesurfer?.setOptions({ waveColor: ERROR_WAVE_COLOR })
			}
		},
		[wavesurfer, onRecordingEnd],
	)

	const handleRetryProcessing = useCallback(async () => {
		dispatch({ type: 'SET_ERROR', payload: '' })
		wavesurfer?.setOptions({ waveColor: DEFAULT_WAVE_COLOR })
		if (recordedBlobRef.current) await handleRecordEnd(recordedBlobRef.current)
	}, [handleRecordEnd, wavesurfer])

	const handleStartRecording = useCallback(async () => {
		dispatch({ type: 'START_RECORDING' })
		isCancelledRef.current = false
		wavesurfer?.empty()
		wavesurfer?.setOptions({
			waveColor: DEFAULT_WAVE_COLOR,
		})
		await wavesurfer?.load('', silentData, 0.5)
		await recordPlugin.current?.startMic()
		await recordPlugin.current?.startRecording()
	}, [wavesurfer])

	const handleStopRecording = useCallback(async () => {
		isCancelledRef.current = false
		recordPlugin.current?.stopRecording()
	}, [])

	const handleCancelRecording = useCallback(async () => {
		if (state.error) {
			// If in a retry state, reset state variables directly
			resetState()
		} else if (state.isRecording) {
			// If recording is in progress, cancel it properly
			isCancelledRef.current = true
			// The 'record-end' event handler will handle resetting state
			recordPlugin.current?.stopRecording()
		}
	}, [state.error, state.isRecording, resetState])

	useEffect(() => {
		const initializeWaveSurfer = async () => {
			if (wavesurfer) {
				const recordPluginInstance = RecordPlugin.create({
					audioBitsPerSecond: 22050,
					scrollingWaveform: true,
					renderRecordedAudio: false,
					scrollingWaveformWindow: 10,
				})

				recordPlugin.current = wavesurfer.registerPlugin(recordPluginInstance)

				recordPlugin.current.on('record-progress', (time: number) => {
					dispatch({ type: 'SET_SECONDS_ELAPSED', payload: time })
				})

				recordPlugin.current.on('destroy', () => {
					recordPlugin.current?.unAll()
				})

				recordPlugin.current.on('record-end', async (blob: Blob) => {
					if (isCancelledRef.current) {
						// Recording was cancelled; do not process blob
						isCancelledRef.current = false // Reset the flag for future recordings
						dispatch({ type: 'STOP_RECORDING' }) // Update the recording state
						dispatch({ type: 'SET_SECONDS_ELAPSED', payload: 0 }) // Reset the timer
						dispatch({ type: 'SET_ERROR', payload: '' }) // Clear any error messages
						wavesurfer?.empty() // Clear the waveform
					} else {
						// Proceed with processing the recorded blob
						recordedBlobRef.current = blob
						await handleRecordEnd(blob)
					}
				})
			}
		}

		void initializeWaveSurfer()

		return () => {
			if (wavesurfer) {
				wavesurfer.destroy()
			}
		}
		// eslint-disable-next-line react-hooks/exhaustive-deps
	}, [wavesurfer])

	return {
		waveformRef,
		state,
		handleStartRecording,
		handleStopRecording,
		handleCancelRecording,
		handleRetryProcessing,
		resetState,
	}
}

const MIN_BAR_HEIGHT = 8 // Small uniform bars for no-sound
const BAR_WIDTH = 2 // Narrow bars
const SPACING = 2 // Small space between bars

// Utility to get pixel ratio for high DPI displays
function getPixelRatio() {
	return Math.max(1, window.devicePixelRatio || 1)
}

// Custom render function to adjust bar height based on amplitude
const renderWaveform = (
	channelData: Array<Float32Array | number[]>,
	ctx: CanvasRenderingContext2D,
) => {
	const options = {
		barWidth: BAR_WIDTH,
		barGap: SPACING,
		barRadius: 0,
		barAlign: 'center',
		barHeight: 2,
	}

	const topChannel = channelData[0]
	if (!topChannel) {
		throw new Error('Top channel data is undefined')
	}
	const bottomChannel = channelData[1] || topChannel
	const length = topChannel.length

	let { width, height } = ctx.canvas
	width += 8
	const halfHeight = height / 2
	const pixelRatio = getPixelRatio()
	const vScale = options.barHeight || 1

	const barWidth = options.barWidth ? options.barWidth * pixelRatio : 1
	const barGap = options.barGap ? options.barGap * pixelRatio : 0
	const barRadius = options.barRadius || 0
	const barIndexScale = width / (barWidth + barGap) / length

	// Clear the canvas for redrawing
	ctx.clearRect(0, 0, width, height)

	ctx.beginPath()

	let prevX = 0
	let maxTop = 0
	let maxBottom = 0

	// Draw the bars
	for (let i = 0; i < length; i++) {
		const x = Math.floor(i * barIndexScale) // Use Math.floor for consistent pixel alignment
		if (x > prevX) {
			const topBarHeight = Math.round(maxTop * halfHeight * vScale)
			const bottomBarHeight = Math.round(maxBottom * halfHeight * vScale)
			let barHeight = topBarHeight + bottomBarHeight || 1

			// Ensure the minimum bar height is at least MIN_BAR_HEIGHT px
			barHeight = Math.max(barHeight, MIN_BAR_HEIGHT)

			// Vertical alignment
			let y = halfHeight - topBarHeight
			if (options.barAlign === 'top') {
				y = 0
			} else if (options.barAlign === 'bottom') {
				y = height - barHeight
			}

			// Draw the bar, fallback to rect if roundRect is not supported
			const rectFn = barRadius && 'roundRect' in ctx ? 'roundRect' : 'rect'
			//@ts-ignore
			ctx[rectFn](
				prevX * (barWidth + barGap),
				barHeight === MIN_BAR_HEIGHT ? halfHeight - 2 : y,
				barWidth,
				barHeight,
				barRadius,
			)

			// Update prevX and reset max values
			prevX = x
			maxTop = 0
			maxBottom = 0
		}

		// Track the max amplitude for the current batch
		const magnitudeTop = Math.abs(topChannel[i] || 0)
		const magnitudeBottom = Math.abs(bottomChannel[i] || 0)
		if (magnitudeTop > maxTop) maxTop = magnitudeTop
		if (magnitudeBottom > maxBottom) maxBottom = magnitudeBottom
	}

	// Finalize the drawing
	ctx.fill()
	ctx.closePath()
}

// Animation loop using requestAnimationFrame
export const renderFunction = (
	channelData: Array<Float32Array | number[]>,
	ctx: CanvasRenderingContext2D,
) => {
	const animate = () => {
		renderWaveform(channelData, ctx)
	}
	requestAnimationFrame(animate) // Sync with the browser's frame ratio
}

const silentData = [
	new Float32Array(500).fill(0.01), // Small values to render minimal bars
	new Float32Array(500).fill(0.01),
]
