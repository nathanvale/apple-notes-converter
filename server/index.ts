/* c8 ignore start */
import 'dotenv/config'
import fs from 'fs'
import { PrismaClient } from '@prisma/client'
import bodyParser from 'body-parser'
import express, { type Request, type Response } from 'express'
import multer from 'multer'
import OpenAI from 'openai'

// Initialize Prisma Client
const prisma = new PrismaClient()

// Initialize OpenAI API with your API key
const openai = new OpenAI({
	apiKey: process.env.OPENAI_API_KEY, // Add your OpenAI API key in the environment variables
})

const app = express()

app.use(express.static('public'))
app.use(bodyParser.json())

// Set up multer for file upload handling
const storage = multer.diskStorage({
	destination: (req, file, cb) => {
		console.log('Saving to /server/uploads/ folder')
		cb(null, './server/uploads')
	},
	filename: (req, file, cb) => {
		const filename = `${Date.now()}-${file.originalname}`
		console.log(`Filename: ${filename}`)
		cb(null, filename)
	},
})

const upload = multer({
	storage,
	limits: { fileSize: 10000000 }, // Limit file size to 10MB for example
	fileFilter: (req, file, cb) => {
		if (file.mimetype !== 'audio/wav' && file.mimetype !== 'audio/mpeg') {
			return cb(new Error('Only .wav or .mp3 files are allowed!'))
		}
		cb(null, true)
	},
})

// Route to handle file upload and transcription
app.post(
	'/upload',
	upload.single('audio'),
	async (req: Request, res: Response) => {
		try {
			const audioFilePath = req.file?.path

			console.log('File uploaded:', req.file) // Log file details

			if (!audioFilePath) {
				return res.status(400).send('No audio file uploaded.')
			}

			// Read the audio file and send it to OpenAI for transcription
			const fileStream = fs.createReadStream(audioFilePath)
			const response = await openai.audio.transcriptions.create({
				file: fileStream,
				model: 'whisper-1',
			})

			const transcription = response.text

			// Save transcription to the database
			const savedTranscription = await prisma.speechTranscription.create({
				data: {
					transcription: transcription,
				},
			})

			// Send back the transcription as the response
			res.json({ transcription: savedTranscription.transcription })

			// Clean up: remove the audio file after processing
			fs.unlink(audioFilePath, (err) => {
				if (err) console.error('Error deleting audio file:', err)
			})
		} catch (error) {
			console.error('Error during transcription:', error)
			res.status(500).send('Error during transcription.')
		}
	},
)

// Start the server
const PORT = process.env.PORT || 3000
app.listen(PORT, () => {
	console.log(`Server is running on port ${PORT}`)
})
