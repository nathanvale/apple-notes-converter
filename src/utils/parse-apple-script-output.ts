export interface AppleScriptNote {
	id: string
	title: string
	content: string
}

export const parseAppleScriptOutput = (output: string): AppleScriptNote[] => {
	try {
		return output
			.split('__________')
			.filter(Boolean)
			.map((noteStr) => {
				if (!noteStr.includes('----------')) {
					throw new Error('Invalid note format')
				}
				const [id, title, content] = noteStr.split('----------')
				const appleScriptNote: AppleScriptNote = {
					id: id?.trim() || '',
					title: title?.trim() || '',
					content: content?.trim() || '',
				}
				return appleScriptNote
			})
	} catch (error) {
		console.error(error)
		throw error
	}
}
