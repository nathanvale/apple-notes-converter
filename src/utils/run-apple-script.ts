import applescript from 'applescript'

export function runAppleScript(script: string): Promise<string> {
	return new Promise((resolve, reject) => {
		applescript.execString(script, (error, result) => {
			if (error) {
				reject(`Error: ${error}`)
			} else {
				resolve(result)
			}
		})
	})
}
