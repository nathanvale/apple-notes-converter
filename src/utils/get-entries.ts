/* c8 ignore start */
;({
	autoLog: false,
})

import { prisma } from './db.server'

async function fetchEntriesWithRelationships() {
	await prisma.entry.findMany({
		include: {
			keyEvents: true, // Include related KeyEvents
			actionItems: true, // Include related ActionItems
		},
	})
}

fetchEntriesWithRelationships()
	.catch((e) => {
		console.error(e)
	})
	.finally(async () => {
		await prisma.$disconnect()
	})
