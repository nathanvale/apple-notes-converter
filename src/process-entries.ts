import { readFile } from 'fs/promises'
import { fileURLToPath } from 'node:url'
import { join } from 'path'
import { type ActionItem, type Entry, type KeyEvent } from '@prisma/client'
import { prisma } from './utils/db.server'
import { parseAppleScriptOutput } from './utils/parse-apple-script-output'
import { parseJournal, type ParsedJournal } from './utils/parse-journal'
import { runAppleScript } from './utils/run-apple-script'

const __filename = fileURLToPath(import.meta.url)
const __dirname = join(__filename, '..') // Get the directory name

async function convertJsonToPrismaModels(jsonOutput: ParsedJournal) {
	const {
		date,
		title,
		journalEntry,
		assessment,
		keyEvents,
		actionItems,
		notes,
	} = jsonOutput

	// Create the Entry first
	const entry: Entry = await prisma.entry.create({
		data: {
			title,
			content: journalEntry,
			date: new Date(date),
			assessment: assessment,
			notes: notes, // Assuming no notes are provided in this case
		},
	})

	// Create the KeyEvents
	const keyEventPromises = keyEvents.map((event: string) =>
		prisma.keyEvent.create({
			data: {
				event,
				entryId: entry.id, // Link to the created Entry
			},
		}),
	)

	const createdKeyEvents: KeyEvent[] = []
	for (const keyEventPromise of keyEventPromises) {
		const createdKeyEvent = await keyEventPromise
		createdKeyEvents.push(createdKeyEvent)
	}

	// Create the ActionItems
	const actionItemPromises = actionItems.map((action: string) =>
		prisma.actionItem.create({
			data: {
				action,
				entryId: entry.id, // Link to the created Entry
			},
		}),
	)

	const createdActionItems: ActionItem[] = []
	for (const actionItemPromise of actionItemPromises) {
		const createdActionItem = await actionItemPromise
		createdActionItems.push(createdActionItem)
	}

	return {
		entry,
		keyEvents: createdKeyEvents,
		actionItems: createdActionItems,
	}
}

export async function processEntries() {
	try {
		const appleScriptPath = join(__dirname, 'get-journal-entries.applescript')
		const appleScript = await readFile(appleScriptPath, 'utf8')

		const output = await runAppleScript(appleScript)

		// Parse the AppleScript output
		const entries = parseAppleScriptOutput(output)

		// Process each entry sequentially
		for (const entry of entries) {
			// Parse the journal entry
			const parsedJournal = parseJournal(entry.content)

			// Convert the parsed journal entry to Prisma models and insert into database
			await convertJsonToPrismaModels(parsedJournal)
		}

		console.log('Entries have been inserted into the database successfully.')

		// Close the Prisma client connection after all operations are done
		await prisma.$disconnect()
	} catch (error) {
		await prisma.$disconnect()
		console.error('An error occurred:', error)
		throw error
	}
}
