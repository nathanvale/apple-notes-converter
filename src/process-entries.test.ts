import { readFile } from 'fs/promises'
import { describe, it, expect, vi, vitest } from 'vitest'
import { readFixture } from '#tests/mocks/utils.js'
import { consoleError } from '#tests/setup/setup-test-env.js'
import { processEntries } from './process-entries'
import { prisma } from './utils/db.server' // Using the actual implementation
import { runAppleScript } from './utils/run-apple-script'
// Mock the required modules except for db.server
vi.mock('fs/promises', () => ({
	readFile: vi.fn(),
}))

vi.mock('./utils/run-apple-script', () => ({
	runAppleScript: vi.fn(),
}))

console.log = vitest.fn()

describe('processEntries', () => {
	it('should process and insert journal entries into the database', async () => {
		const mockAppleScript = 'mock AppleScript content'
		const mockOutput = await readFixture('entries', '1').then(
			(data) => data.output,
		)
		const mockEntries = [
			{
				id: 1,
				title: '2024/08/24 Title 3',
				content:
					'2024/08/24 Title 3\n## Key Events\n- Had lots of energy despite a big night\n- Spent meaningful time with Melanie\n\n## Action Items\n- Prioritize good sleep to recover from the weekend\n- Continue nurturing the growing bond between Melanie and Levi\n\n## Journal Entry\n\nToday, I was surprised by how much energy I had, despite the big night before.\n\n## Psychological Assessment\n\nToday was a day of balancing energy and emotions.',
			},
			{
				id: 2,
				title: '2024/08/25 Title 2',
				content:
					'2024/08/25 Title 2\n\n## Key Events\n\t•\tWorked on parsing journal entries for database import\n\t•\tSpent quality time with Levi at a scout event and Queen Vic Market\n\n## Action Items\n\n\t•\tContinue refining the journal entry parsing process to ensure seamless database integration.\n\t•\tPlan more structured activities with Levi to help manage the challenges of extended time together.\n\n\n## Journal Entry\n\nToday, I achieved a significant milestone by figuring out how to parse all my journal entries so they could be imported into a database. \n\n## Psychological Assessment\n\nToday was a day of positive progress and personal reflection.',
			},
			{
				id: 3,
				title: '2024/08/26 Title 1',
				content:
					'202 4/08/26 Title 1\n\n## Key Events\n- Got on top of personal tasks, including work-related and Scout-related responsibilities.\n- Felt proud of completing various tasks and clearing backlog.\n\n## Action Items\n- Continue to prioritize time for personal space and self-care.\n- Maintain intentionality in time spent with Melanie.\n\n## Journal Entry\n\nToday was a significant day in terms of productivity and personal growth.\n\n\n## Psychological Assessment\n\nToday was marked by a strong sense of accomplishment and self-awareness.',
			},
		]

		// Mock readFile to return mock AppleScript content
		vi.mocked(readFile).mockResolvedValue(mockAppleScript)

		// Mock runAppleScript to return mock AppleScript output
		vi.mocked(runAppleScript).mockResolvedValue(mockOutput)

		// Run the function
		await processEntries()

		// Verify that the entries were inserted into the database
		const entriesInDb = await prisma.entry.findMany()
		expect(entriesInDb).toHaveLength(mockEntries.length)
		expect(entriesInDb).toMatchSnapshot()
	})

	it('should handle errors from read file and close the Prisma client connection', async () => {
		consoleError.mockImplementation(() => {})
		const mockError = new Error('Mock error')

		vi.mocked(readFile).mockRejectedValue(mockError)

		const disconnectSpy = vi.spyOn(prisma, '$disconnect')

		await expect(processEntries()).rejects.toThrow('Mock error')
		expect(disconnectSpy).toHaveBeenCalled()
	})

	it('should handle errors from runAppleScript and close the Prisma client connection', async () => {
		consoleError.mockImplementation(() => {})
		const mockAppleScript = 'mock AppleScript content'
		const mockError = new Error('Mock error')

		vi.mocked(readFile).mockResolvedValue(mockAppleScript)
		vi.mocked(runAppleScript).mockRejectedValue(mockError)

		const disconnectSpy = vi.spyOn(prisma, '$disconnect')

		await expect(processEntries()).rejects.toThrow('Mock error')
		expect(disconnectSpy).toHaveBeenCalled()
	})

	it('should close the Prisma client connection after successful operation', async () => {
		const mockAppleScript = 'mock AppleScript content'
		const mockOutput = await readFixture('entries', '1').then(
			(data) => data.output,
		)

		vi.mocked(readFile).mockResolvedValue(mockAppleScript)
		vi.mocked(runAppleScript).mockResolvedValue(mockOutput)

		const disconnectSpy = vi.spyOn(prisma, '$disconnect')

		await processEntries()

		expect(disconnectSpy).toHaveBeenCalled()
	})
})
