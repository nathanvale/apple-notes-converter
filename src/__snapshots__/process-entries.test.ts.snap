// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`processEntries > should process and insert journal entries into the database 1`] = `
[
  {
    "actionItems": [
      {
        "action": "Prioritize good sleep to recover from the weekend",
        "entryId": 1,
        "id": 1,
      },
      {
        "action": "Continue nurturing the growing bond between Melanie and Levi",
        "entryId": 1,
        "id": 2,
      },
    ],
    "assessment": "Today was a day of balancing energy and emotions.",
    "content": "Today, I was surprised by how much energy I had, despite the big night before.",
    "date": 2024-08-23T14:00:00.000Z,
    "id": 1,
    "keyEvents": [
      {
        "entryId": 1,
        "event": "Had lots of energy despite a big night",
        "id": 1,
      },
      {
        "entryId": 1,
        "event": "Spent meaningful time with Melanie",
        "id": 2,
      },
    ],
    "notes": null,
    "title": "Title 3",
  },
  {
    "actionItems": [
      {
        "action": "Continue refining the journal entry parsing process to ensure seamless database integration.",
        "entryId": 2,
        "id": 3,
      },
      {
        "action": "Plan more structured activities with Levi to help manage the challenges of extended time together.",
        "entryId": 2,
        "id": 4,
      },
    ],
    "assessment": "Today was a day of positive progress and personal reflection.",
    "content": "Today, I achieved a significant milestone by figuring out how to parse all my journal entries so they could be imported into a database.",
    "date": 2024-08-24T14:00:00.000Z,
    "id": 2,
    "keyEvents": [
      {
        "entryId": 2,
        "event": "Worked on parsing journal entries for database import",
        "id": 3,
      },
      {
        "entryId": 2,
        "event": "Spent quality time with Levi at a scout event and Queen Vic Market",
        "id": 4,
      },
    ],
    "notes": null,
    "title": "Title 2",
  },
  {
    "actionItems": [
      {
        "action": "Continue to prioritize time for personal space and self-care.",
        "entryId": 3,
        "id": 5,
      },
      {
        "action": "Maintain intentionality in time spent with Melanie.",
        "entryId": 3,
        "id": 6,
      },
    ],
    "assessment": "Today was marked by a strong sense of accomplishment and self-awareness.",
    "content": "Today was a significant day in terms of productivity and personal growth.",
    "date": 2024-08-25T14:00:00.000Z,
    "id": 3,
    "keyEvents": [
      {
        "entryId": 3,
        "event": "Got on top of personal tasks, including work-related and Scout-related responsibilities.",
        "id": 5,
      },
      {
        "entryId": 3,
        "event": "Felt proud of completing various tasks and clearing backlog.",
        "id": 6,
      },
    ],
    "notes": null,
    "title": "Title 1",
  },
]
`;
